templates/install_puppet.ps1.erb

```diff
@@ -0,0 +1,185 @@
+      Move-Item -Path "$InstallDir\puppet\bin\puppetres.dll" -Destination $temp_puppetres
+    }
+    return $temp_puppetres
+  }
+}
+
+<#
+.Synopsis
+  Restore puppetres.dll to the original location. This should only be used when an installation fails
+.Parameter temp_puppetres
+  Location of the temporary puppetres.dll file.
+#>
+function Script:restore_puppetres_dll {
+  param(
+    [Parameter(Mandatory=$true)]
+    [String] $temp_puppetres
+  )
+  begin {
+    # _Never_ use the $InstallDir top-level parameter to try and find the InstallDir
+    # for the puppetres workaround. The workaround should _always_ fetch the InstallDir
+    # from the registry. This is so a user can specify a different InstallDir than the
+    # directory where the current package is installed and the workaround will still work
+    # for the already installed package
+    $InstallDir = fetch_install_dir
+    if (Test-Path $temp_puppetres) {
+      Out-File -InputObject "Restoring puppetres.dll" -FilePath $Logfile -Append
+      Move-Item -Path $temp_puppetres -Destination "$InstallDir\puppet\bin\puppetres.dll"
+    }
+  }
+}
+
+# ************** Execution start **************
+$ErrorActionPreference = "Stop"
+$service_names=@(
+  "puppet",
+  "pxp-agent",
+  "mcollective"
+)
+$msi_arguments = "/qn /norestart /i `"$Source`" /l*vx+ `"$Logfile`" $InstallArgs"
+
+if (Test-Path "HKLM:\SOFTWARE\Puppet Labs") {
+  $msi_arguments += " ENVIRONMENT=`"$(puppet.bat config print --section agent environment)`""
+}
+if ($InstallDir) {
+  $msi_arguments += " INSTALLDIR=`"$InstallDir`""
+}
+if ($PuppetMaster) {
+  $msi_arguments += " PUPPET_MASTER_SERVER=`"$PuppetMaster`""
+}
+if ($PuppetStartType) {
+  $msi_arguments += " PUPPET_AGENT_STARTUP_MODE=`"$PuppetStartType`""
+}
+Out-File -InputObject $PID -FilePath $InstallScriptPIDFile
+try {
+  # Wait for the puppet run to finish
+  Out-File -InputObject "Waiting for puppet to stop, PID:$PuppetPID" -FilePath $Logfile -Append
+  try {
+    Wait-Process -Id $PuppetPID -Timeout 120
+  # We have to catch all exceptions here since the Error action preference of 'stop' will cause wait-process
+  # to actually throw [System.Management.Automation.ActionPreferenceStopException] rather than the actual
+  # exceptions. However, once inside the catch block the value of $_ is correctly set as the actual exception
+  } catch {
+    if ($_.Exception.GetType() -eq [Microsoft.PowerShell.Commands.ProcessCommandException]) {
+      Out-File -InputObject "Puppet was already finished, continuing..." -FilePath $Logfile -Append
+    } elseif ($_.Exception.GetType() -eq [System.TimeoutException]) {
+      Out-File -InputObject "ERROR: timed out waiting for puppet run to finish!" -FilePath $Logfile -Append
+      throw $_
+    } else {
+      Out-File -InputObject "ERROR: failed while trying to wait for puppet process with $($_.Exception.GetType())" -FilePath $Logfile -Append
+      throw $_
+    }
+  }
+  foreach($service in $service_names) {

```

Comment:
The previous `install_puppet.bat` file had a long comment about why stopping the services was important. Maybe that comment should be here too?